README.md
# **LFQ - lockfree queue**

The current implementation of a **lock-free queue** in C is based on the paper
and sample code of Ruslan Nikolaev. See https://github.com/rusnikola/lfqueue
for details. The paper describes various different methods.

The code of lpq is based on the scqd method and assembled, so that it can
be easily used in your own project, without adding new libraries. See the **example.c** code for its usage.

A lot has been written about lock-free queues and their implementations.
It's a rather tricky topic to get it right. Many still lack a proper
implementation.

## lfq usage:

Copy the files in the **src** directory into your project folder. 
Include the header in your project:

`#include <scqd.h>`

`queue_t *lfq_init(size_t order);`

Initialize a new queue. The size of the queue is 2^order. Returns: 
Pointer to new queue.  

`void *lfq_enqueue(queue_t *q, void *val);`  

Puts the object *val into the queue. Returns:  
**LFQ_OK** if successful  
**LFQ_FULL** if unsuccessful  
**LFQ_CLOSED** if queue is closed.  

`void *lfq_dequeue(queue_t *q);`

Gets an object from the queue. Returns:  
The object, if successful  
**LFQ_EMPTY** if no objects are available  
**LFQ_CLOSED** if queue is closed.  

`void lfq_close(queue_t *q);`

Closes the queue. No more objects can be put in the queue.
Remaining objects can be retrieved, until the queue is empty.

`void lfq_free(queue_t *q);`

Frees the memory from the queue.



ChangeLog
2020-05-23
- Fix calloc size in queue_init
- Fix compile issue on OpenBSD
- Fix queue closed test in dequeue
- Initial Release 1.0


src/lf.h
/*
  Copyright (c) 2018,2019, Ruslan Nikolaev
  All rights reserved.
*/

/* vi: set tabstop=4: */

#ifndef __BITS_LF_H
#define __BITS_LF_H	1

#include <inttypes.h>
#include <sys/types.h>
#include <stdbool.h>
#include <stddef.h>


// PH: included LFCONGIG_H

/* For the following architectures, it is cheaper to use split (word-atomic)
   loads whenever possible. */
#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) ||	\
	defined(__aarch64__)
# define __LFLOAD_SPLIT(dtype_width)	(dtype_width > LFATOMIC_WIDTH)
#else
# define __LFLOAD_SPLIT(dtype_width)	0
#endif

/* IA-64 provides a 128-bit single-compare/double-swap instruction, so
   LFCMPXCHG_SPLIT is true for 128-bit types. */
#if defined(__ia64__)
# define __LFCMPXCHG_SPLIT(dtype_width)	(dtype_width > LFATOMIC_WIDTH)
#else
# define __LFCMPXCHG_SPLIT(dtype_width)	0
#endif

#if defined(__x86_64__) || defined (__aarch64__) || defined(__powerpc64__)	\
	|| (defined(__mips__) && _MIPS_SIM == _MIPS_SIM_ABI64)
typedef uint64_t lfepoch_t;
typedef int64_t lfepoch_signed_t;
typedef int64_t lfsatomic_t;
typedef uint64_t lfatomic_t;
typedef __uint128_t lfatomic_big_t;
# define LFATOMIC_LOG2			3
# define LFATOMIC_WIDTH			64
# define LFATOMIC_BIG_WIDTH		128
#elif defined(__i386__) || defined(__arm__) || defined(__powerpc__)			\
	|| (defined(__mips__) &&												\
		(_MIPS_SIM == _MIPS_SIM_ABI32 || _MIPS_SIM == _MIPS_SIM_NABI32))
# if defined(__i386__)
typedef uint64_t lfepoch_t; /* Still makes sense to use 64-bit numbers. */
typedef int64_t lfepoch_signed_t;
# else
typedef uint32_t lfepoch_t;
typedef int32_t lfepoch_signed_t;
# endif
typedef int32_t lfsatomic_t;
typedef uint32_t lfatomic_t;
typedef uint64_t lfatomic_big_t;
# define LFATOMIC_LOG2			2
# define LFATOMIC_WIDTH			32
# define LFATOMIC_BIG_WIDTH		64
#else
typedef uintptr_t lfepoch_t;
typedef intptr_t lfsatomic_t;
typedef uintptr_t lfatomic_t;
typedef uintptr_t lfatomic_big_t;
# if UINTPTR_MAX == UINT32_C(0xFFFFFFFF)
#  define LFATOMIC_LOG2			2
#  define LFATOMIC_WIDTH		32
#  define LFATOMIC_BIG_WIDTH	32
# elif UINTPTR_MAX == UINT64_C(0xFFFFFFFFFFFFFFFF)
#  define LFATOMIC_LOG2			3
#  define LFATOMIC_WIDTH		64
#  define LFATOMIC_BIG_WIDTH	64
# endif
#endif

/* XXX: True for x86/x86-64 but needs to be properly defined for other CPUs. */
#define LF_CACHE_SHIFT		7U
#define LF_CACHE_BYTES		(1U << LF_CACHE_SHIFT)

/* Allow to use LEA for x86/x86-64. */
#if defined(__i386__) || defined(__x86_64__)
# define __LFMERGE(x,y)	((x) + (y))
#else
# define __LFMERGE(x,y)	((x) | (y))
#endif

#ifdef __GNUC__
# define __LF_ASSUME(c) do { if (!(c)) __builtin_unreachable(); } while (0)
#else
# define __LF_ASSUME(c)
#endif

/* GCC does not have a sane implementation of wide atomics for x86-64
   in recent versions, so use inline assembly workarounds whenever possible.
   No aarch64 support in GCC for right now. */
#if (defined(__i386__) || defined(__x86_64__)) && defined(__GNUC__) &&	\
	!defined(__llvm__) && defined(__GCC_ASM_FLAG_OUTPUTS__)
# include "gcc_x86.h"
#else
# include "c11.h"
#endif

/* ABA tagging with split (word-atomic) load/cmpxchg operation. */
#if __LFLOAD_SPLIT(LFATOMIC_BIG_WIDTH) == 1 ||								\
		__LFCMPXCHG_SPLIT(LFATOMIC_BIG_WIDTH) == 1
# define __LFABA_IMPL(w, type_t)											\
static const size_t __lfaba_shift##w = sizeof(lfatomic_big_t) * 4;			\
static const size_t __lfaptr_shift##w = 0;									\
static const lfatomic_big_t __lfaba_mask##w =								\
				~(lfatomic_big_t) 0U << (sizeof(lfatomic_big_t) * 4);		\
static const lfatomic_big_t __lfaba_step##w =								\
				(lfatomic_big_t) 1U << (sizeof(lfatomic_big_t) * 4);
#endif

/* ABA tagging when load/cmpxchg is not split. Note that unlike previous
   case, __lfaptr_shift is required to be 0. */
#if __LFLOAD_SPLIT(LFATOMIC_BIG_WIDTH) == 0 &&								\
		__LFCMPXCHG_SPLIT(LFATOMIC_BIG_WIDTH) == 0
# define __LFABA_IMPL(w, type_t)											\
static const size_t __lfaba_shift##w = sizeof(type_t) * 8;					\
static const size_t __lfaptr_shift##w = 0;									\
static const lfatomic_big_t __lfaba_mask##w =								\
				~(lfatomic_big_t) 0U << (sizeof(type_t) * 8);				\
static const lfatomic_big_t __lfaba_step##w =								\
				(lfatomic_big_t) 1U << (sizeof(type_t) * 8);
#endif

typedef bool (*lf_check_t) (void * data, void * addr, size_t size);

static inline size_t lf_pow2(size_t order)
{
	return (size_t) 1U << order;
}

static inline bool LF_DONTCHECK(void * head, void * addr, size_t size)
{
	return true;
}

#define lf_container_of(addr, type, field)									\
	(type *) ((char *) (addr) - offsetof(type, field))

#ifdef __cplusplus
# define LF_ERROR	UINTPTR_MAX
#else
# define LF_ERROR	((void *) UINTPTR_MAX)
#endif

/* Available on all 64-bit and CAS2 32-bit architectures. */
#if LFATOMIC_BIG_WIDTH >= 64
__LFABA_IMPL(32, uint32_t)
#endif

/* Available on CAS2 64-bit architectures. */
#if LFATOMIC_BIG_WIDTH >= 128
__LFABA_IMPL(64, uint64_t)
#endif

/* Available on CAS2 32/64-bit architectures. */
#if LFATOMIC_BIG_WIDTH >= 2 * __LFPTR_WIDTH
__LFABA_IMPL(, uintptr_t)
#endif

#endif	/* !__BITS_LF_H */

/* vi: set tabstop=4: */


src/c11.h
/*
  Copyright (c) 2019, Ruslan Nikolaev
  All rights reserved.
*/

#ifndef __LF_C11_H
#define __LF_C11_H 1

#include <stdatomic.h>

#define LFATOMIC(x)				_Atomic(x)
#define LFATOMIC_VAR_INIT(x)	ATOMIC_VAR_INIT(x)

static inline void __lfaba_init(_Atomic(lfatomic_big_t) * obj,
		lfatomic_big_t val)
{
	atomic_init(obj, val);
}

static inline lfatomic_big_t __lfaba_load(_Atomic(lfatomic_big_t) * obj,
		memory_order order)
{
#if __LFLOAD_SPLIT(LFATOMIC_BIG_WIDTH) == 1
	lfatomic_big_t res;
	_Atomic(lfatomic_t) * hobj = (_Atomic(lfatomic_t) *) obj;
	lfatomic_t * hres = (lfatomic_t *) &res;

	hres[0] = atomic_load_explicit(hobj, order);
	hres[1] = atomic_load_explicit(hobj + 1, order);
	return res;
#elif __LFLOAD_SPLIT(LFATOMIC_BIG_WIDTH) == 0
	return atomic_load_explicit(obj, order);
#endif
}

static inline bool __lfaba_cmpxchg_weak(_Atomic(lfatomic_big_t) * obj,
	lfatomic_big_t * expected, lfatomic_big_t desired,
	memory_order succ, memory_order fail)
{
	return atomic_compare_exchange_weak_explicit(obj, expected, desired,
					succ, fail);
}

static inline bool __lfaba_cmpxchg_strong(_Atomic(lfatomic_big_t) * obj,
	lfatomic_big_t * expected, lfatomic_big_t desired,
	memory_order succ, memory_order fail)
{
	return atomic_compare_exchange_strong_explicit(obj, expected, desired,
					succ, fail);
}

/* XXX remove
static inline dtype_t __lfaba_fetch_and(_Atomic(lfatomic_big_t) * obj,
		lfatomic_big_t arg, memory_order order)
{
	return atomic_fetch_and_explicit(obj, arg, order);
}
*/

#endif /* !__LF_C11_H */

/* vi: set tabstop=4: */


src/scqd.h
/*-
 * Copyright (c) 2020 Peter Haag
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef SCQD_H
#define SCQD_H

#define LFRING_EMPTY	(~(size_t) 0U)

#define LFQ_OK	   (void *) ((void *)1)
#define LFQ_CLOSED (void *) ((void *)-2)
#define LFQ_FULL   (void *) ((void *)-3)
#define LFQ_EMPTY  (void *) LFRING_EMPTY


// Generic queue type - content encapsulated
typedef void queue_t;

/* 
 * Init the queue
 *    order: queue size = 2^order
 */
queue_t *lfq_init(size_t order);

void *lfq_enqueue(queue_t *q, void *val);

void *lfq_dequeue(queue_t *q);

void lfq_close(queue_t *q);

void lfq_free(queue_t *q);

void lfq_dump(queue_t *q);

#endif 


example/example.c
/*-
 * Copyright (c) 2020 Peter Haag
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <stdint.h>
#include <pthread.h>

#include "scqd.h"

const int max = 10000000;

static int total;
static pthread_mutex_t lock; 

static void *consumer(void *_queue);

static void *consumer(void *_queue){
queue_t *queue = (queue_t *)_queue;

	uintptr_t cnt = 0;
	int empty_cnt = 0;
	do {
		void *data = lfq_dequeue(queue);
		if ( data == LFQ_CLOSED ) {
			printf("Q closed\n");
			break;
		}
		if ( data == LFQ_EMPTY ) {
			empty_cnt++;
		} else {
			cnt++;
		}
	} while (1);

	pthread_mutex_lock(&lock);
	total += cnt;
	pthread_mutex_unlock(&lock);

	printf("thread done: %lu, empty_cnt: %d\n", cnt, empty_cnt);
	pthread_exit(NULL);
}

int main(int argc, char **argv) {

	queue_t *queue;

	total = 0;
	pthread_mutex_init(&lock, NULL);
	queue = lfq_init(15);

	pthread_t _thread[2];
	pthread_create(&_thread[0],NULL,consumer,queue);
	pthread_create(&_thread[1],NULL,consumer,queue);

	uintptr_t cnt = 1;
	uintptr_t failed_cnt = 0;
	do {
		if ( lfq_enqueue(queue, (void *)cnt) != LFQ_OK ) {
			failed_cnt++;
			printf("enqueue empty_cnt\n");
		} else {
			cnt++;
		}
	} while ( cnt <= max );
	cnt--;

	lfq_close(queue);
	printf("Feed done: %lu, failed enqueues: %lu\n", cnt, failed_cnt);

	pthread_join(_thread[0],NULL);
	pthread_join(_thread[1],NULL);

	printf("Done. Total consumed count: %d\n", total);
	lfq_free(queue);

	return 0;
}


src/scqd.c
/*-
 * Copyright (c) 2019 Ruslan Nikolaev.  All Rights Reserved.
 * Copyright (c) 2020 Peter Haag
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/* vi: set tabstop=4: */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <inttypes.h>
#include <sys/types.h>
#include <stdatomic.h>
#include <string.h>

#include "lf.h"
#include "scqd.h"

#if LFATOMIC_WIDTH == 32
# define LFRING_MIN	(LF_CACHE_SHIFT - 2)
#elif LFATOMIC_WIDTH == 64
# define LFRING_MIN	(LF_CACHE_SHIFT - 3)
#elif LFATOMIC_WIDTH == 128
# define LFRING_MIN	(LF_CACHE_SHIFT - 4)
#else
# error "Unsupported LFATOMIC_WIDTH."
#endif

struct __lfring {
	__attribute__ ((aligned(LF_CACHE_BYTES))) _Atomic(lfatomic_t) head;
	__attribute__ ((aligned(LF_CACHE_BYTES))) _Atomic(lfsatomic_t) threshold;
	__attribute__ ((aligned(LF_CACHE_BYTES))) _Atomic(lfatomic_t) tail;
	__attribute__ ((aligned(LF_CACHE_BYTES))) _Atomic(lfatomic_t) array[1];
};

typedef struct _queue_s {
	__attribute__ ((aligned(LF_CACHE_BYTES))) _Atomic(lfatomic_t) closed;
	size_t order;
	struct __lfring *aq;
	struct __lfring *fq;
	void **val;
} _queue_t;

#define LFRING_ALIGN	(_Alignof(struct __lfring))
#define LFRING_SIZE(o)	\
	(offsetof(struct __lfring, array) + (sizeof(lfatomic_t) << ((o) + 1)))


#define __lfring_cmp(x, op, y)	((lfsatomic_t) ((x) - (y)) op 0)

#if LFRING_MIN != 0
static inline size_t __lfring_map(lfatomic_t idx, size_t order, size_t n) {
	return (size_t) (((idx & (n - 1)) >> (order + 1 - LFRING_MIN)) |
			((idx << LFRING_MIN) & (n - 1)));
}
#else
static inline size_t __lfring_map(lfatomic_t idx, size_t order, size_t n) {
	return (size_t) (idx & (n - 1));
}
#endif

#define __lfring_threshold3(half, n) ((long) ((half) + (n) - 1))

static inline size_t lfring_pow2(size_t order) {
	return (size_t) 1U << order;
}

static inline void lfring_init_empty(struct __lfring *q, size_t order) {
	size_t i, n = lfring_pow2(order + 1);

	for (i = 0; i != n; i++)
		atomic_init(&q->array[i], (lfsatomic_t) -1);

	atomic_init(&q->head, 0);
	atomic_init(&q->threshold, -1);
	atomic_init(&q->tail, 0);
}

static inline void lfring_init_fill(struct __lfring *q, size_t s, size_t e, size_t order) {
	size_t i, half = lfring_pow2(order), n = half * 2;

	for (i = 0; i != s; i++)
		atomic_init(&q->array[__lfring_map(i, order, n)], 2 * n - 1);
	for (; i != e; i++)
		atomic_init(&q->array[__lfring_map(i, order, n)], n + i);
	for (; i != n; i++)
		atomic_init(&q->array[__lfring_map(i, order, n)], (lfsatomic_t) -1);

	atomic_init(&q->head, s);
	atomic_init(&q->threshold, __lfring_threshold3(half, n));
	atomic_init(&q->tail, e);
}

static inline bool lfring_enqueue(struct __lfring *q, size_t order, size_t eidx, bool nonempty) {
	size_t tidx, half = lfring_pow2(order), n = half * 2;
	lfatomic_t tail, entry, ecycle, tcycle;

	eidx ^= (n - 1);

	while (1) {
		tail = atomic_fetch_add_explicit(&q->tail, 1, memory_order_acq_rel);
		tcycle = (tail << 1) | (2 * n - 1);
		tidx = __lfring_map(tail, order, n);
		entry = atomic_load_explicit(&q->array[tidx], memory_order_acquire);
retry:
		ecycle = entry | (2 * n - 1);
		if (__lfring_cmp(ecycle, <, tcycle) && ((entry == ecycle) ||
				((entry == (ecycle ^ n)) &&
				 __lfring_cmp(atomic_load_explicit(&q->head,
				  memory_order_acquire), <=, tail)))) {

			if (!atomic_compare_exchange_weak_explicit(&q->array[tidx],
					&entry, tcycle ^ eidx,
					memory_order_acq_rel, memory_order_acquire))
				goto retry;

			if (!nonempty && (atomic_load(&q->threshold) != __lfring_threshold3(half, n)))
				atomic_store(&q->threshold, __lfring_threshold3(half, n));
			return true;
		}
	}
}

static inline void __lfring_catchup(struct __lfring *q, lfatomic_t tail, lfatomic_t head) {
	while (!atomic_compare_exchange_weak_explicit(&q->tail, &tail, head,
			memory_order_acq_rel, memory_order_acquire)) {
		head = atomic_load(&q->head);
		tail = atomic_load(&q->tail);
		if (__lfring_cmp(tail, >=, head))
			break;
	}
}

static inline size_t lfring_dequeue(struct __lfring *q, size_t order, bool nonempty) {
	size_t hidx, n = lfring_pow2(order + 1);
	lfatomic_t head, entry, entry_new, ecycle, hcycle, tail;
	size_t attempt;

	if (!nonempty && atomic_load(&q->threshold) < 0) {
		return LFRING_EMPTY;
	}

	while (1) {
		head = atomic_fetch_add_explicit(&q->head, 1, memory_order_acq_rel);
		hcycle = (head << 1) | (2 * n - 1);
		hidx = __lfring_map(head, order, n);
		attempt = 0;
again:
		entry = atomic_load_explicit(&q->array[hidx], memory_order_acquire);

		do {
			ecycle = entry | (2 * n - 1);
			if (ecycle == hcycle) {
				atomic_fetch_or_explicit(&q->array[hidx], (n - 1),
						memory_order_acq_rel);
				return (size_t) (entry & (n - 1));
			}

			if ((entry | n) != ecycle) {
				entry_new = entry & ~(lfatomic_t) n;
				if (entry == entry_new)
					break;
			} else {
				if (++attempt <= 10000)
					goto again;
				entry_new = hcycle ^ ((~entry) & n);
			}
		} while (__lfring_cmp(ecycle, <, hcycle) &&
					!atomic_compare_exchange_weak_explicit(&q->array[hidx],
					&entry, entry_new,
					memory_order_acq_rel, memory_order_acquire));

		if (!nonempty) {
			tail = atomic_load_explicit(&q->tail, memory_order_acquire);
			if (__lfring_cmp(tail, <=, head + 1)) {
				__lfring_catchup(q, tail, head + 1);
				atomic_fetch_sub_explicit(&q->threshold, 1,
					memory_order_acq_rel);
				return LFRING_EMPTY;
			}

			if (atomic_fetch_sub_explicit(&q->threshold, 1,
					memory_order_acq_rel) <= 0)
				return LFRING_EMPTY;
		}
	}
}

queue_t *lfq_init(size_t order) {

	_queue_t *queue = calloc(1, sizeof(_queue_t));
	if ( !queue )
		return NULL;

	// int queue struct
	queue->order = order;
	atomic_init(&queue->closed, 0);

	// allocate one chunk of memory depending on order
	size_t totalSize = 2 * (LFRING_SIZE(order)) + (1U << order) * sizeof(void *);
	void *queueMem = calloc(1, totalSize);
	if ( !queueMem ) {
		free(queue);
		return NULL;
	}

	// assign memory to queue
	queue->aq  = queueMem;
	queue->fq  = (struct __lfring *)(queueMem + LFRING_SIZE(order));
	queue->val = (void **)(queueMem + 2*(LFRING_SIZE(order)));

	// init rings
	lfring_init_empty(queue->aq, order);
	lfring_init_fill(queue->fq, 0, 1U << order, order);

	return (queue_t *)queue;

} // End of lfq_init

void *lfq_enqueue(queue_t *q, void * val) {
	_queue_t *queue = (_queue_t *)q;

	if ( atomic_load(&queue->closed) )
		return LFQ_CLOSED;

	size_t eidx;
	eidx = lfring_dequeue(queue->fq, queue->order, true);
	if (eidx == LFRING_EMPTY)
		return LFQ_FULL;

	queue->val[eidx] = val;

	lfring_enqueue(queue->aq, queue->order, eidx, false);
	return LFQ_OK;

} // End of lfq_enqueue

void *lfq_dequeue(queue_t * q) {
	_queue_t *queue = (_queue_t *)q;

	size_t eidx = lfring_dequeue(queue->aq, queue->order, false);
	if (eidx == LFRING_EMPTY) {
		if ( atomic_load(&queue->closed) )
			return LFQ_CLOSED;
		return LFQ_EMPTY;
	}

	void *val = queue->val[eidx];
	lfring_enqueue(queue->fq, queue->order, eidx, true);

	return val;
} // End of lfq_dequeue

void lfq_close(queue_t * q) {
	_queue_t *queue = (_queue_t *)q;
	
	atomic_fetch_add_explicit(&queue->closed, 1, memory_order_acq_rel);
	return;
} // End of lfq_close

void lfq_free(queue_t *q) {
	_queue_t *queue = (_queue_t *)q;

	free(queue->aq);
	free(queue);
} // End of lfq_free

void lfq_dump(queue_t *q) {
	_queue_t *queue = (_queue_t *)q;

	for (int i = 0; i < (1U << queue->order); i++)
		printf("slot: %u - %lu\n", i, (uintptr_t)queue->val[i]);

} // End of lfq_dump


